packages/text-input/src/TextInput.tsx

```diff
@@ -0,0 +1,331 @@
+ * @param props.placeholder The placeholder text shown in the input field before the user begins typing.
+ * @param props.errorMessage The message shown below the input field if the value is invalid.
+ * @param props.state The current state of the TextInput. This can be none, valid, or error.
+ * @param props.value The current value of the input field. If a value is passed to this prop, component will be controlled by consumer.
+ * @param props.setValue Callback used to set the value of the input field.
+ * @param props.className className supplied to the TextInput container.
+ */
+const TextInput = React.forwardRef(
+  (
+    {
+      label,
+      description,
+      optional = false,
+      disabled = false,
+      onChange,
+      placeholder,
+      errorMessage,
+      state,
+      value: controlledValue,
+      setValue: setControlledValue,
+      className,
+    }: TextInputProps,
+    forwardRef: React.Ref<HTMLInputElement>,
+  ) => {
+    const isControlled = typeof controlledValue === 'string';
+    const [uncontrolledValue, setUncontrolledValue] = useState('');
+    const value = isControlled ? controlledValue : uncontrolledValue;
+    const setValue = isControlled ? setControlledValue : setUncontrolledValue;
+    const { usingKeyboard: showFocus } = useUsingKeyboardContext();
+    const [hasFocus, setHasFocus] = useState(false);
+    const inputSelectorProp = createDataProp('input-selector');
+
+    function getInputColorFromState() {
+      if (state === State.None) {
+        return '#CCC';
+      }
+
+      return state === State.Valid ? uiColors.green.base : uiColors.red.base;
+    }
+
+    function getInputPaddingFromState() {
+      if (state === State.Valid || state === State.Error) {
+        return '30px';
+      }
+
+      return optional ? '60px' : '12px';
+    }
+
+    function onValueChange(e: React.ChangeEvent<HTMLInputElement>) {
+      if (onChange !== undefined) {
+        state = onChange(e);
+      }
+
+      if (setValue != undefined) {
+        setValue(e.target.value);
+      }
+    }
+
+    const interactionRingFocusStyle = css`
+      ${inputSelectorProp.selector}:focus ~ & {
+        background-color: #9dd0e7;
+        transform: scale(1);
+        z-index: 1;
+      }
+    `;
+
+    const interactionRingHoverStyle = css`
+      ${inputSelectorProp.selector}:hover ~ & {
+        transform: scale(1);
+      }
+    `;
+
+    const descriptionStyle = css`
+      color: ${uiColors.gray.dark1};
+      font-size: 14px;
+      height: ${description === '' ? '0' : '20'}px;
+      font-weight: normal;
+      padding-bottom: 8px;
+    `;
+
+    const conditionalStyling = css`
+      z-index: ${hasFocus ? 2 : 1};
+      background: ${disabled ? uiColors.gray.light2 : '#FFFFFF'};
+      padding-right: ${getInputPaddingFromState()};
+      border: 1px solid ${getInputColorFromState()};
+    `;
+
+    return (
+      <label className={cx(textInputStyle, labelStyle, className)}>
+        {label}
```

Comment:
This should be inside the label element instead.

Did you know about a `label`'s `for` attribute? It's what lets a user click on a label and have it act on the input element. [Docs here](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/label)

We'll need to create a unique id for each TextInput but we should definitely have a `for` attribute to lean into the default browser behavior.

Reply:
I've seen the `for` attribute mentioned before, but I didn't touch it because I wasn't sure how to handle the unique ID piece. How would I use that attribute here? Also- isn't {label} already inside the element? I'm not sure what you mean.